src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSObject.cs

```diff
@@ -16,7 +16,16 @@ public partial class JSObject : IDisposable
         /// <summary>
         /// Returns true if the proxy was already disposed.
         /// </summary>
-        public bool IsDisposed { get => _isDisposed; }
+        public bool IsDisposed
+        {
+            get
+            {
+                lock (ProxyContext)
```

Comment:
Other code already appears to acquire a lock before reading .IsDisposed of a JSObject, we should always do it in one place or the other.

Reply:
also i should note that you could end up with a deadlock due to this if other code incorrectly does `lock (a) { ... = obj.IsDisposed; }` and `obj.ProxyContext` is a different context that is currently locked by another worker

reading the IsDisposed flag without the protection of a lock is a data race, so i don't think we need to defend against it by acquiring the lock in the property getter this way. i think we can make it the caller's responsibility

Reply:
I disagree. Since there is no logic inside of `IsDisposed` getter, it could not lead to deadlock, because it's not ever locking another lock inside.

I do this to memory fence. There are too many places this is called from otherwise unlocked `ObjectDisposedException.ThrowIf(self.IsDisposed, self)` to make it caller's responsibility.

Reply:
ok, I changed my mind and do as you say :)
Nested lock should be cheap.